exchanges/bitfinex/bitfinex_websocket.go

```diff
@@ -1521,15 +1521,56 @@ func (b *Bitfinex) WsUpdateOrderbook(p currency.Pair, assetType asset.Item, book
 				err)
 		}
 
-		err = validateCRC32(ob, checkme.Token)
-		if err != nil {
+		if err = validateCRC32(ob, checkme.Token); err != nil {
+			log.Errorf(log.WebsocketMgr, "%s websocket orderbook update error, will resubscribe orderbook: %v", b.Name, err)
+			if suberr := b.resubOrderbook(p, assetType); suberr != nil {
+				log.Errorf(log.ExchangeSys, "%s error resubscribing orderbook: %v", b.Name, suberr)
+			}
 			return err
 		}
 	}
 
 	return b.Websocket.Orderbook.Update(&orderbookUpdate)
 }
 
+// resubOrderbook resubscribes the orderbook after a consistency error, probably a failed checksum,
+// which forces a fresh snapshot. If we don't do this the orderbook will keep erroring and drifting.
+func (b *Bitfinex) resubOrderbook(p currency.Pair, assetType asset.Item) error {
+	if err := b.Websocket.Orderbook.FlushOrderbook(p, assetType); err != nil {
+		return err
+	}
+
+	c, err := b.chanForSub(wsBook, assetType, p)
+	if err != nil {
+		return err
+	}
+	return b.Websocket.ResubscribeToChannel(c)
+}
+
+// chanForSub returns an existing channel subscription for a given channel/asset/pair
+func (b *Bitfinex) chanForSub(cName string, assetType asset.Item, pair currency.Pair) (*stream.ChannelSubscription, error) { //nolint:unparam // cName open for future uses
+	var c *stream.ChannelSubscription
+	want := &stream.ChannelSubscription{
+		Channel:  cName,
+		Currency: pair,
+		Asset:    assetType,
+	}
+	subs := b.Websocket.GetSubscriptions()
+	for i := range subs {
+		if subs[i].Equal(want) {
+			if c != nil {
+				return nil, errTooManyMatchingSubs
+			}
+			c = &subs[i]
```

Comment:
I would return &subs[i] straight away. If we are subbed to 3k items for example it will need to check all items even if found which can be quite costly even though this is only for one websocket orderbook resub. It should really be checked when adding to the subscriptions instead.

Reply:
Let's leave it as-is, if that's okay?
My bigger solution will do away with this, but I don't want to fold that into this because the scopes barely overlap.

Reply:
Actually, no. Let's do a good job now.
Okay.

Reply:
Done.